pkg/controller/core/workload_controller.go

```diff
@@ -103,15 +143,39 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 				"Inadmissible", fmt.Sprintf("ClusterQueue %s is inactive", cqName))
 			return ctrl.Result{}, client.IgnoreNotFound(err)
 		}
-	case admitted:
-		msg := fmt.Sprintf("Admitted by ClusterQueue %s", wl.Spec.Admission.ClusterQueue)
-		err := workload.UpdateStatusIfChanged(ctx, r.client, &wl, kueue.WorkloadAdmitted, metav1.ConditionTrue, "AdmissionByKueue", msg)
+	case cancellingAdmission:
+		err := workload.UpdateStatusIfChanged(ctx, r.client, &wl, kueue.WorkloadAdmitted, metav1.ConditionFalse,
+			"AdmissionCancelled", "Admission cancelled")
 		return ctrl.Result{}, client.IgnoreNotFound(err)
+	case admitted:
+		if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadAdmitted) {
+			return r.reconcileNotReadyTimeout(ctx, req, &wl)
+		} else {
+			msg := fmt.Sprintf("Admitted by ClusterQueue %s", wl.Spec.Admission.ClusterQueue)
+			err := workload.UpdateStatusIfChanged(ctx, r.client, &wl, kueue.WorkloadAdmitted, metav1.ConditionTrue, "AdmissionByKueue", msg)
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
 	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *WorkloadReconciler) reconcileNotReadyTimeout(ctx context.Context, req ctrl.Request, wl *kueue.Workload) (ctrl.Result, error) {
+	countingTowardsTimeout, recheckAfter := r.admittedNotReadyWorkload(wl, realClock)
+	if !countingTowardsTimeout {
+		return ctrl.Result{}, nil
+	}
+	if recheckAfter > 0 {
+		klog.V(4).InfoS("Workload not yet ready and did not exceed its timeout", "workload", req.NamespacedName.String(), "recheckAfter", recheckAfter)
+		return ctrl.Result{RequeueAfter: recheckAfter}, nil
+	} else {
+		klog.V(2).InfoS("Cancelling admission of the workload due to exceeding the PodsReady timeout", "workload", req.NamespacedName.String())
+		wl.Spec.Admission = nil
```

Comment:
Don't we need to clone first? I can't remember if the controller-runtime client does a copy.

Although I have a different proposal for how to patch admission that doesn't require cloning https://github.com/kubernetes-sigs/kueue/pull/514/files#diff-a7490e7f5efa2a59ab508f592b67dd6e7530b9d2e1b19a1b6a7438e6c2386905

WDYT?

Reply:
I think that updating the workload in-place without a copy was working ok, because a new instance is created at the beginning of `Reconcile` and it is not used after cancelling admission. 

However, I agree it is better to avoid in-place modification and I like the approach you suggested so I've applied it here too.